models/auth/oidc.go

```diff
@@ -0,0 +1,255 @@
+		cfg.OidcTokenEndpointUrl = tmpStr
+	}
+
+	return cfg
+}
+
+type oidcServerConfig struct {
+	// OidcIssuer specifies the issuer to verify OIDC tokens with. This issuer
+	// will be used to load public keys to verify signature and will be compared
+	// with the issuer claim in the OIDC token. It will be used if
+	// AuthenticationMethod == "oidc". By default, this value is "".
+	OidcIssuer string `json:"oidc_issuer"`
+	// OidcAudience specifies the audience OIDC tokens should contain when validated.
+	// If this value is empty, audience ("client ID") verification will be skipped.
+	// It will be used when AuthenticationMethod == "oidc". By default, this
+	// value is "".
+	OidcAudience string `json:"oidc_audience"`
+	// OidcSkipExpiryCheck specifies whether to skip checking if the OIDC token is
+	// expired. It will be used when AuthenticationMethod == "oidc". By default, this
+	// value is false.
+	OidcSkipExpiryCheck bool `json:"oidc_skip_expiry_check"`
+	// OidcSkipIssuerCheck specifies whether to skip checking if the OIDC token's
+	// issuer claim matches the issuer specified in OidcIssuer. It will be used when
+	// AuthenticationMethod == "oidc". By default, this value is false.
+	OidcSkipIssuerCheck bool `json:"oidc_skip_issuer_check"`
+}
+
+func getDefaultOidcServerConf() oidcServerConfig {
+	return oidcServerConfig{
+		OidcIssuer:          "",
+		OidcAudience:        "",
+		OidcSkipExpiryCheck: false,
+		OidcSkipIssuerCheck: false,
+	}
+}
+
+func unmarshalOidcServerConfFromIni(conf ini.File) oidcServerConfig {
+	var (
+		tmpStr string
+		ok     bool
+	)
+
+	cfg := getDefaultOidcServerConf()
+
+	if tmpStr, ok = conf.Get("common", "oidc_issuer"); ok {
+		cfg.OidcIssuer = tmpStr
+	}
+
+	if tmpStr, ok = conf.Get("common", "oidc_audience"); ok {
+		cfg.OidcAudience = tmpStr
+	}
+
+	if tmpStr, ok = conf.Get("common", "oidc_skip_expiry_check"); ok && tmpStr == "true" {
+		cfg.OidcSkipExpiryCheck = true
+	} else {
+		cfg.OidcSkipExpiryCheck = false
+	}
+
+	if tmpStr, ok = conf.Get("common", "oidc_skip_issuer_check"); ok && tmpStr == "true" {
+		cfg.OidcSkipIssuerCheck = true
+	} else {
+		cfg.OidcSkipIssuerCheck = false
+	}
+
+	return cfg
+}
+
+type OidcAuthProvider struct {
+	baseConfig
+
+	tokenGenerator *clientcredentials.Config
+}
+
+func NewOidcAuthSetter(baseCfg baseConfig, cfg oidcClientConfig) *OidcAuthProvider {
+	tokenGenerator := &clientcredentials.Config{
+		ClientID:     cfg.OidcClientId,
+		ClientSecret: cfg.OidcClientSecret,
+		Scopes:       []string{cfg.OidcAudience},
+		TokenURL:     cfg.OidcTokenEndpointUrl,
+	}
+
+	return &OidcAuthProvider{
+		baseConfig:     baseCfg,
+		tokenGenerator: tokenGenerator,
+	}
+}
+
+func (auth *OidcAuthProvider) generateAccessToken() (accessToken string, err error) {
+	tokenObj, err := auth.tokenGenerator.Token(context.Background())
+	if err != nil {
+		return "", fmt.Errorf("couldn't generate OIDC token for login: %v", err)
+	}
+	return tokenObj.AccessToken, nil
+}
```

Comment:
@GuyLewin I contact oauth2 and oidc recently, but meet some problem when I used oidc auth. 

From the code, frpc use `access_token` as `PrivilegeKey` when login operation, and then frps will verify the  `PrivilegeKey`. 

```
func (v *IDTokenVerifier) Verify(ctx context.Context, rawIDToken string)
```

My confusion: why use "IDToken" to verify "access_token". As far as I know, `access_token` is from Oauth2 provider's response, and `id_token` is from OIDC response.  It seems that `access_token`is uncorrelated to `id_token`. 

I maybe have a misunderstanding, but can you help me?

Reply:
Hey @yuyulei , sorry for the delayed response.
You're right that sometimes access_token isn't JWT, and that won't be supported by this library.
We needed to use `access_token` since we wanted to verify OIDC permissions as well to a specific resource, not only the identity of the login. Perhaps it's wrong to rely on access_token being JWT, what is the access_token type you're getting?

Reply:
Thanks for your response, @GuyLewin . 

At first, I found some issues about oidc, so I tried AliyunCloud as third-party `oauth` provider,  but failed. It seems that the `access_token` of AliyunCloud is not generated by `JWT`. 

Frp'users may choose different `oauth` provider. We are not enable to guarantee these cloud providers use `JWT` to generate `access_token`. Usually in `oauth` protocal, We send `access_token` to providers, and they will help us to verify it. Is it right ?

So how about use `id_token` to verify authentication ? By `oidc` protocal, we get `access_token` and `id_token` from `oidc` providers, then we can decode `id_token` by `JWT`. I think this way will be more common, just as `oidc` protocal.

Reply:
@yuyulei as I mentioned in my previous comment, using id_token won't allow you to verify the user actually has permissions to access the FRP server. I think it'll only allow you to know who the user is